twins/nats/pubsub.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package nats
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/twins"
+	"github.com/mainflux/mainflux/twins/mqtt"
+	"github.com/mainflux/senml"
+	"github.com/nats-io/go-nats"
+)
+
+const (
+	queue = "twins"
+	input = "channel.>"
+)
+
+var mqttOp = map[string]string{
+	"stateSucc": "state/success",
+	"stateFail": "state/failure",
+}
+
+type pubsub struct {
+	natsClient *nats.Conn
+	mqttClient mqtt.Mqtt
+	logger     log.Logger
+	twins      twins.TwinRepository
+	states     twins.StateRepository
+}
+
+// Subscribe to appropriate NATS topic
+func Subscribe(nc *nats.Conn, mc mqtt.Mqtt, tr twins.TwinRepository, sr twins.StateRepository, logger log.Logger) {
+	ps := pubsub{
+		natsClient: nc,
+		mqttClient: mc,
+		logger:     logger,
+		twins:      tr,
+		states:     sr,
+	}
+	ps.natsClient.QueueSubscribe(input, queue, ps.handleMsg)
+}
+
+func (ps pubsub) handleMsg(m *nats.Msg) {
+	var msg mainflux.Message
+	err := proto.Unmarshal(m.Data, &msg)
+	if err != nil {
```

Comment:
You can inline this.

Reply:
No, I need this err defined for defer func (see below) to work.